### ISSUE TITLE
"invalid disk name" when cloning from a template and specifying a disk name

### ISSUE THREAD

**Issue by jason-azze** [terraform-providers/terraform-provider-vsphere#301]
### Terraform Version

Terraform v0.11.1
+ provider.vsphere v1.1.0

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
resource "vsphere_virtual_machine" "generic" {
  name             = "${var.name_prefix}${format("%02d", count.index + var.name_starting_val)}"
  resource_pool_id = "${data.vsphere_resource_pool.target-resource-pool.id}"
  datastore_id     = "${data.vsphere_datastore.target-datastore.id}"

  count                = "${var.how_many}"
  folder               = "${var.vm_folder_name}"
  num_cpus             = "${var.vcpu_count}"
  num_cores_per_socket = "${var.cores_per_socket}"
  memory               = "${var.memory_amount}"
  guest_id             = "${data.vsphere_virtual_machine.source-template.guest_id}"
  scsi_type            = "${data.vsphere_virtual_machine.source-template.scsi_type}"

   disk {
    name             = "${var.name_prefix}${format("%02d", count.index + var.name_starting_val)}disk01.vmdk"
    size             = "${data.vsphere_virtual_machine.source-template.disks.0.size}"
    thin_provisioned = "${data.vsphere_virtual_machine.source-template.disks.0.thin_provisioned}"
  }

```

### Expected Behavior
A new VM with a disk with the filename (in this particular case) cent69-build04disk01.vmdk should be created.

I believe (perhaps incorrectly) that `name =` inside the disk block in the case of a clone operation is the name of the vmdk to be created. I can put whatever I want here. "hotdogsaredelicious.vmdk"

If I _must_ use the VM name as the `name =` of the disk, then what would I use as the `name =` of a second or third disk?

### Actual Behavior
```
module.linux-build.vsphere_virtual_machine.generic: disk.0: invalid disk name "cent69-build04disk01.vmdk" for cloning. Please rename this disk to "cent69-build04.vmdk"
*module.utility-agents.vsphere_virtual_machine.generic: 1 error(s) occurred:
```
### Steps to Reproduce

1. `terraform plan`

### Important Factoids
Rich Siegel over on Gitter pointed me to the code and accompanying comment that is probably generating this error message.
https://github.com/terraform-providers/terraform-provider-vsphere/blob/master/vsphere/internal/virtualdevice/virtual_machine_disk_subresource.go#L604-L616
But, I don't grasp what the "existing disk" is in this case. There is only the source VM (in this case a template) and the VM to be provisioned, which doesn't exist yet, so doesn't have an "existing disk".

EDITS: Moving too fast on a Friday afternoon. Fixed a bunch of typos and unclear sections.

**Comment by vancluever** [comment]
Hey @jason-azze, you are correct in your assessment that this is the disk that gets created, but that's only half the story.

vSphere clones the disk during a clone with the general default naming way that a VM gets during creation. This follows the convention of `VMNAME.vmdk` for the initial disk, and `VMNAME_INDEX.vmdk` for all other disks, hence `foobar.vmdk`, `foobar_1.vmdk`, `foobar_2.vmdk`, and so on.

You have no control over this naming during a clone - in the relocate spec the fields in the backing are ignored. See [here](https://pubs.vmware.com/vsphere-6-5/topic/com.vmware.wssdk.apiref.doc/vim.vm.RelocateSpec.DiskLocator.html):

> Specific property changes may be ignored if they are not supported.
> 
> Supported BackingInfo types and properties:
> 
> VirtualDiskFlatVer2BackingInfo
>   thinProvisioned
>   eagerlyScrub
> VirtualDiskSeSparseBackingInfo (ESX 5.1 or later)

Hence we have to put restrictions on what the devices are named in the event of clones to ensure they can be properly tracked. This is why this needs to be restricted.

Note that this only applies to the disks that come from the template. Any disks that you create after the fact can bear whatever name you want.

So in summary, your first disk should bear the name of the VM, and then from so on, it should be the name, an underscore, and then the index.

Hope this helps!

**Comment by jason-azze** [comment]
Ahh. It's strange, but I understand your explanation and the constraints. It's like the Ford Model T, you can have any color you want as long as it's black.

Thanks for the help @vancluever .
